### PR TITLE
feat(bitbucket-server): Support git over ssh

### LINKED ISSUE #6280: [bitbucket-server] git clone over ssh
**What would you like Renovate to be able to do?**
gitFs over ssh for bitbucket-server

**Describe the solution you'd like**
The same as for other platforms

**Describe alternatives you've considered**
no

**Additional context**
Our self hosted bitbucket-server instance has disabled HTTPS cloning. 
There is no way we can use https for cloning. It would be great if you supported this request. Thanks.

### PR DESCRIPTION
**Pull request by danez**
This allows checkout of repositories from bitbucket-server when http access for git is not enabled

## Changes:

- in `initRepo()` I moved the API calls to be done before the actual git clone, so we can read the clone urls from the API response.
- Use urls from API response or fallback to current URL generation
  - Try to use http URL from API with injected auth
  - If http not available try to use ssh url
  - Otherwise generate URL from endpoint
- Removed test `does not throw` as it was exactly the same as `works`
- Added tests for all cases and adjusted the repoMock to be able to configure the clone URLs in the mock response.

## Context:

Fixes #6280 

This allows to use bitbucket-server instances where https is turned of for git:

Screenshot 2020-12-22 at 12 19 51

## Documentation (please check one with an [x])

`TBD: Still missing in self-hosted config, Want to figure out how to handle known_hosts for now. See below`

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [x] Unit tests + ran on a real repository

I did run real tests with local docker version of bb version 6.0.11 and 7.8.1.

**_One important thing_**: The ssh access did NOT work out of the box, because the host verification was not stored in `known_hosts`. I assume this is not limited to bb-server but also happens with gitlab for example. How is that handled with gitlab? Is this up to the user to add the identification to `known_hosts` on the machine where renovate is running? I couldn't find anything in the docs about this. I see it was mentioned here https://github.com/renovatebot/renovate/issues/4022#issuecomment-508129259 but not sure what the outcome was.

### CODE REVIEW

**Review by monstro805 — APPROVED** [review]

File: lib/platform/bitbucket-server/index.ts [hunk 1]
```diff
@@ -43,6 +43,18 @@ import { smartTruncate } from '../utils/pr-body';
...
+  parent: string;
+  links: {
+    clone: { href: string; name: string }[];
+  };
+}
+
+interface BranchResponse {
+  displayId: string;
+}
+
```
Comment (viceice) [inline lib/platform/bitbucket-server/index.ts:57]: Move this to `./types.ts`

File: lib/platform/bitbucket-server/index.ts [hunk 2]
```diff
@@ -210,6 +232,14 @@ export async function initRepo({
...
     throw err;
   }
+
+  config.mergeMethod = 'merge';
+  const repoConfig: RepoResult = {
+    defaultBranch: branchRes.body.displayId,
+    isFork: !!info.parent,
+  };
+
+  return repoConfig;
```
Comment (viceice) [inline lib/platform/bitbucket-server/index.ts:242]: move this to inner try block

File: lib/platform/bitbucket-server/index.ts [hunk 3]
```diff
@@ -154,37 +166,17 @@ export async function initRepo({
...
-    ...config,
-    localDir,
-    url: gitUrl,
-    gitAuthorName: global.gitAuthor?.name,
-    gitAuthorEmail: global.gitAuthor?.email,
-    cloneSubmodules,
-  });
-
+  let info: RepoResponse;
+  let branchRes: HttpResponse<BranchResponse>;
```
Comment (viceice) [inline lib/platform/bitbucket-server/index.ts:170]: move this to inner try block

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by zeldigas** [comment]
@danez any chance you have time to address leftover comments? Looks like PR is close to merge :)

**Comment by danez** [comment]
> @danez any chance you have time to address leftover comments? Looks like PR is close to merge :)

I will

File: lib/platform/bitbucket-server/index.ts (hunk 2 above)
Comment (danez) [reply lib/platform/bitbucket-server/index.ts]: done

File: lib/platform/bitbucket-server/index.ts (hunk 1 above)
Comment (danez) [reply lib/platform/bitbucket-server/index.ts]: done, also renamed on type that was named `Bbbs` whereas all other types are named `BbsX`

**Comment by zeldigas** [comment]
@viceice, @rarkins could you take a look?

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
Has this been able to be tested on both types of BBS - https URLs and ssh?

**Review by viceice — APPROVED** [review]

**Comment by danez** [comment]
> Has this been able to be tested on both types of BBS - https URLs and ssh?

I did with a docker image of bbs version 6.0 and 7.8 locally.
As mentioned above the ssh did not work out of the box, and i had to first add the bbs instance to the ssh known_hosts file first in the container where renovate did run. But otherwise it did work.

**Comment by zeldigas** [comment]
Thanks a ton!

**Pull request merged**